### PR TITLE
Split Batcher logic with OpenGl implementation

### PR DESCRIPTION
**Pull request by karupayun**
Following decisions in go/stadia-orbit-testing-ui-elements, we want to
refactor Batcher functionality to make it easier to mock. In this PR we
are creating OpenGlBatcher (first step in http://b/225173189).

In addition we cleaned includes and refactor virtual methods to not have
default parameters.

Next step: Create PrimitiveAssembler.

### CODE REVIEW

File: src/OrbitGl/BatcherImpl.h
Comment (beckerhe) [inline src/OrbitGl/BatcherImpl.h:81]: After seeing the code, I could also imagine having multi level inheritance. Something like that:

- `Batcher` (inherits from nothing) declares a bunch of pure virtual functions for adding primitives
- `DrawingBatcher` (inherits from `Batcher`) declares `GetLayers` and `DrawLayer` as pure virtual functions
- `DrawingBatcherImpl` (inherits from `DrawingBatcher`) implements all of it. This class could be declared in a cpp-file and be a proper implementation detail.

For your tests you can mock or fake `Batcher`. The `CaptureWindow` would get a pointer to `DrawingBatcher` because it needs `DrawLayer`, but all the rest probably only needs a pointer to `Batcher`.

**Review by beckerhe — COMMENTED** [review]
Since you're already changing the names, would you consider moving Batcher and BatcherInterface into the proper namespace?

File: src/OrbitGl/Batcher.h
Comment (reichlfl) [inline src/OrbitGl/Batcher.h:47]: ... that **have** to be implemented ....

File: src/OrbitGl/BatcherImpl.h
Comment (reichlfl) [reply src/OrbitGl/BatcherImpl.h:81]: Additional thoughts: You mention you want to rename this to `BatcherInterface`, but if I understand correctly, this will still include implementation of non-private methods? In that case it's not an interface. Did you consider composition over inheritance? For example:

* `BatcherInterface` declares only the virtual functions needed to add primitives, but as public API
* `OpenGlBatcher` implements this interface
* Create a new class `PrimitiveAssembler` (... better names welcome) that receives a pointer to a `BatcherInterface` and is responsible of assembling complex elements (circles, shaded boxes, ...) using the `BatcherInterface` methods

**Review by antonrohr — COMMENTED** [review]
Thanks for including me on this PR. I do not have comments right now, but maybe after you addressed the other comments. 

Regarding:
> Since you're already changing the names, would you consider moving Batcher and BatcherInterface into the proper namespace?

I think this is a good idea and your upcoming renaming PR is a good place for that

File: src/OrbitGl/BatcherImpl.h
Comment (karupayun) [reply src/OrbitGl/BatcherImpl.h:81]: Really thanks for the suggestions. A few thoughts. I feel:

- The goal of this PR is isolate OpenGl related methods with the rest in Batcher (Thanks Florian for the better name, I should use OpenGlBatcher!).
- Henning is trying to additionally isolate drawing primitives with adding them.
- Florian is trying to additionally isolate proper primitives with additional methods to add them. 

I found both proposals correct, I was a bit lazy to think about all this refactors in this PR, but I should have a clearly defined long-time proposal before merging this, thanks for making me realize that.

If we are worried about the long-term picture, I will include merging TextRenderer and Batcher. For example, in Florian's proposal, PrimitiveAssembler could have both pointer of BatcherInterface and TextRendererInterface.

I will update the doc before going forward with this PR.

Comment (reichlfl) [reply src/OrbitGl/BatcherImpl.h:81]: Fair enough to not do this all in a single PR, this can absolutely be approached iteratively and I think this is the first step in the right direction.

**Comment by beckerhe** [comment]
> Thanks for including me on this PR. I do not have comments right now, but maybe after you addressed the other comments.
> 
> Regarding:
> 
> > Since you're already changing the names, would you consider moving Batcher and BatcherInterface into the proper namespace?
> 
> I think this is a good idea and your upcoming renaming PR is a good place for that

Oh, I think I misunderstood that. Are you planning to do the renaming in a separate PR?

File: src/OrbitGl/BatcherImpl.h
Comment (karupayun) [reply src/OrbitGl/BatcherImpl.h:67]: > Since you're already changing the names, would you consider moving Batcher and BatcherInterface into the proper namespace?

Will add it for OpenGlBatcher. After changing Batcher to BatcherInterface I will also include it in orbit_gl namespace.

**Comment by karupayun** [comment]
> Since you're already changing the names, would you consider moving Batcher and BatcherInterface into the proper namespace?

Will add it for OpenGlBatcher. After changing Batcher to BatcherInterface I will also include it in orbit_gl namespace.

**Comment by karupayun** [comment]
Create a bug with steps: http://b/225173189.

Renamed to OpenGlBatcher, added a few namespace, change a comment. 

Asking for a review now that we have a clearer path.

**Review by antonrohr — APPROVED** [review]
LGTM, a second review other than mine might be a good idea, I might have missed something

File: src/OrbitGl/OpenGlBatcher.h
Comment (beckerhe) [inline src/OrbitGl/OpenGlBatcher.h:80]: Nit: We tend to use `//` instead of `/* */`

Comment (beckerhe) [reply src/OrbitGl/OpenGlBatcher.h:80]: Also an empty line above the comment would be great

File: src/OrbitGl/OpenGlBatcher.h
Comment (beckerhe) [inline src/OrbitGl/OpenGlBatcher.h:83]: `explicit`

File: src/OrbitGl/OpenGlBatcher.h
Comment (beckerhe) [inline src/OrbitGl/OpenGlBatcher.h:89]: Why does this need to be `protected`? Couldn't this be `private`?

File: src/OrbitGl/OpenGlBatcher.h
Comment (beckerhe) [inline src/OrbitGl/OpenGlBatcher.h:20]: You shouldn't have an anonymous namespace in a header. If you want to have these types internal put it into the namespace `orbit_gl_internal`.

**Review by beckerhe — APPROVED** [review]
Otherwise LGTM

**Review by reichlfl — APPROVED** [review]
Please address Henning's comments, otherwise LGTM!

File: src/OrbitGl/OpenGlBatcher.h
Comment (karupayun) [reply src/OrbitGl/OpenGlBatcher.h:89]: It's used internally in batcher test. Do you prefer me to provide a protected method to get it?

File: src/OrbitGl/OpenGlBatcher.h
Comment (karupayun) [reply src/OrbitGl/OpenGlBatcher.h:80]: Yes, it was copy pasted directly from Batcher.

**Pull request merged**
